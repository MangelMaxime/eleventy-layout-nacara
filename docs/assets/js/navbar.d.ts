declare function setupNavbarBurger(): void;
declare function setupMobileMenu(): void;
declare function setupNavbarDropdown(): void;
