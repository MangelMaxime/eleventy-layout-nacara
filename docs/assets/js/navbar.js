const setupNavbarBurger = () => {
    /*
     * Setup menu burger behaviour
     */

    // Code copied from Bulma documentation
    // https://bulma.io/documentation/components/navbar/#navbar-menu

    const $navbarBurgerDots = document.querySelector(".navbar-burger-dots");

    if ($navbarBurgerDots !== null) {
        $navbarBurgerDots.addEventListener("click", (ev) => {
            const $nacaraNavbarMenu = document.querySelector(
                ".nacara-navbar-menu"
            );

            if ($nacaraNavbarMenu !== null) {
                $nacaraNavbarMenu.classList.toggle("is-active");
                $navbarBurgerDots.classList.toggle("is-active");
            }
        });
    }
};
console.log("dizhduizhdiz")
const setupMobileMenu = () => {
    const mobileMenuTrigger = document.querySelector(
        ".mobile-menu .menu-trigger"
    );

    if (mobileMenuTrigger !== null) {
        mobileMenuTrigger.addEventListener("click", () => {
            document
                .querySelector(".is-menu-column")
                .classList.toggle("force-show");

            mobileMenuTrigger.classList.toggle("is-active");
        });
    }
};

const setupNavbarDropdown = () => {
    // Generate a unique data-id attribute for each dropdown
    // We do it at runtime to keep the HTML output clean
    // If user re-generate a page without changing the content the HTML output will be the same
    let nextDropdownId = 0;

    document
        .querySelectorAll(".navbar-item.has-nacara-dropdown")
        .forEach(function (navbarItem) {
            navbarItem.setAttribute("data-id", nextDropdownId);
            nextDropdownId++;
        });

    document
        .querySelectorAll(".navbar-item.has-nacara-dropdown")
        .forEach(function (navbarItem) {
            navbarItem.addEventListener("click", (ev) => {
                // Click is inside the dropdown element, nothing to do
                if (ev.target.closest(".nacara-dropdown")) {
                    return;
                }

                const $activeDropdownItem = document.querySelector(
                    ".navbar-item.has-nacara-dropdown.is-active"
                );

                // If we clicked on the active dropdown close it
                if ($activeDropdownItem) {
                    const activeItemId =
                        $activeDropdownItem.attributes.getNamedItem(
                            "data-id"
                        ).value;
                    const clickedItemId =
                        navbarItem.attributes.getNamedItem("data-id").value;

                    // If user clicked on the active dropdown item, close everything
                    if (activeItemId === clickedItemId) {
                        // Close the dropdown
                        $activeDropdownItem.classList.remove("is-active");
                        // Remove the overlay
                        $greyOverlay.classList.remove("is-active");
                    } else {
                        // The user clicked on another dropdown item, close the active one
                        $activeDropdownItem.classList.remove("is-active");
                        // Show the new dropdown as active
                        navbarItem.classList.add("is-active");
                    }
                } else {
                    // Show the overlay
                    $greyOverlay.classList.add("is-active");
                    // Show which dropdown is active
                    navbarItem.classList.add("is-active");
                }
            });
        });

    const navbarDropdownTimeoutCache = {};

    // Fullwidth dropdown needs to handle the hover event via JavaScript to add a small delay
    // before hiding the dropdown when the mouse leave
    // This is because when displaying a dropdown in fullwidth there is a gap between
    // the dropdown-link and the dropdown content
    document
        .querySelectorAll(".navbar-item.has-nacara-dropdown.is-fullwidth")
        .forEach(function (navbarItem) {
            navbarItem.addEventListener("mouseenter", function () {
                const itemGuid =
                    navbarItem.attributes.getNamedItem("data-id").value;

                // If there is already a timeout for this dropdown cancel it
                if (navbarDropdownTimeoutCache[itemGuid]) {
                    clearTimeout(navbarDropdownTimeoutCache[itemGuid]);
                    delete navbarDropdownTimeoutCache[itemGuid];
                }

                // Add the CSS class which force to show the dropdown from the JS
                navbarItem.classList.add("is-hovered-js");

                // When the mouse leave, wait a small amount of time before removing the CSS class
                navbarItem.addEventListener(
                    "mouseleave",
                    function (ev) {
                        const timeoutId = setTimeout(() => {
                            navbarItem.classList.remove("is-hovered-js");
                            delete navbarDropdownTimeoutCache[itemGuid];
                        }, 200);
                        // Store the timeoutID so we can cancel it if the user make some back and forth
                        // between the link and the dropdown content
                        navbarDropdownTimeoutCache[itemGuid] = timeoutId;
                    },
                    { once: true }
                );
            });
        });

    const $greyOverlay = document.querySelector(".grey-overlay");

    if ($greyOverlay !== null) {
        $greyOverlay.addEventListener("click", (ev) => {
            const $activeDropdownItem = document.querySelector(
                ".navbar-item.has-nacara-dropdown.is-active"
            );

            if ($activeDropdownItem) {
                $activeDropdownItem.classList.remove("is-active");
                $greyOverlay.classList.remove("is-active");
            }
        });
    }
};

// The page is ready to execute our code
if (document.readyState === "complete") {
    setupNavbarBurger();
    setupNavbarDropdown();
    setupMobileMenu();
    // The page is not ready, wait for it to be ready
} else {
    document.onreadystatechange = () => {
        if (document.readyState === "complete") {
            setupNavbarBurger();
            setupNavbarDropdown();
            setupMobileMenu();
        }
    };
}
