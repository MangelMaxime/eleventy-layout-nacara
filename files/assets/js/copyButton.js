// Setup the copy code button for snippets
const setupCopyCode = () => {
    const snippetElements = Array.from(document.querySelectorAll("pre > code"));

    snippetElements
        .forEach(codeElement => {
            // If one of the parent of codeElement has data-disable-copy-button attributes
            // do not had the copy button
            // We store this information on a parent because we don't control the `snippet` generation
            if (codeElement.closest("[data-disable-copy-button]")) {
                return;
            }

            const copyButton = document.createElement("button");
            copyButton.innerText = "Copy";
            copyButton.classList.add(
                "button",
                "is-primary",
                "is-outlined",
                "is-copy-button"
            );

            const codeText = codeElement.innerText;

            copyButton.addEventListener("click", () => {
                // Copy the code into the clipboard
                const $input = document.createElement("textarea");
                document.body.appendChild($input);
                $input.value = codeText;
                $input.select();
                document.execCommand("copy");
                $input.remove();

                // Notify the user
                copyButton.innerText = "Copied";
                // Revert the button text
                window.setTimeout(() => {
                    copyButton.innerText = "Copy";
                }, 1000)
            })

            codeElement.appendChild(copyButton);
        });
}

// The page is ready to execute our code
if (document.readyState === "complete") {
    setupCopyCode();
    // The page is not ready, wait for it to be ready
} else {
    document.onreadystatechange = () => {
        if (document.readyState === "complete") {
            setupCopyCode();
        }
    }
}
