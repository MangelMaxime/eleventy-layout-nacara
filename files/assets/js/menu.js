const scrollMenuOrTableOfContentIfNeeded = () => {

    // Make the table of content visible
    const $tableOfContentElement = document.querySelector(".table-of-content");
    const $activeMenuItemElement = document.querySelector(".menu .is-active");

    if ($tableOfContentElement !== null) {
        scrollIntoView($tableOfContentElement, {
            scrollMode: "if-needed",
            block: 'nearest',
            inline: 'nearest',
            boundary: document.querySelector(".menu-container")
        });
    } else if ($activeMenuItemElement !== null) {
        scrollIntoView($activeMenuItemElement, {
            scrollMode: "if-needed",
            block: 'nearest',
            inline: 'nearest',
            boundary: document.querySelector(".menu-container")
        });
    }
}

// The page is ready to execute our code
if (document.readyState === "complete") {
    scrollMenuOrTableOfContentIfNeeded();
    // The page is not ready, wait for it to be ready
} else {
    document.onreadystatechange = () => {
        if (document.readyState === "complete") {
            scrollMenuOrTableOfContentIfNeeded();
        }
    }
}
