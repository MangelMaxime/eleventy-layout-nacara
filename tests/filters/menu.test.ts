import { expect, test } from 'vitest'
import { formatHTML } from "../utils/_formatHTML.js";
import Eleventy from "@11ty/eleventy";

test("returns nothing if there is not nacaraMenu provided", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-0/", "./tests/fixtures/menu-0/_site", {
        configPath: "./tests/fixtures/menu-0/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});

test("supports nacaraMenu with a Section", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-1/", "./tests/fixtures/menu-1/_site", {
        configPath: "./tests/fixtures/menu-1/.eleventy.js",
    });

    const json = await elev.toJSON() as any [];
    const page1Json = json.find(
        (item: any) => item.url === "/docs/getting-started/page1/"
    );
    const page1FormattedResult = await formatHTML(page1Json.content);
    const page2Json = json.find(
        (item: any) => item.url === "/docs/getting-started/page2/"
    );
    const page2FormattedResult = await formatHTML(page2Json.content);

    expect(page1FormattedResult).toMatchSnapshot();
    expect(page2FormattedResult).toMatchSnapshot();
});

// Disabled because it pollutes the console output too much...
// I don't understand why I can make it work using normal expect / toThrow API
// test("crash if the menu contains nested sections", async () => {
//     const elev = new Eleventy("./tests/fixtures/menu-2/", "./tests/fixtures/menu-2/_site", {
//         configPath: "./tests/fixtures/menu-2/.eleventy.js",
//     });
//     try {
//         await elev.toJSON() as any[];
//     } catch (e: any) {
//         expect(e.originalError.message).toContain(`(./fixtures/menu-2/_includes/menu.njk)
//   Nested sections are not supported by eleventy-layout-nacara`)
//     }
// });

test("returns a menu with a TOC containing only h2 headers if no toc data is provided", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-3/", "./tests/fixtures/menu-3/_site", {
        configPath: "./tests/fixtures/menu-3/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});

test("returns a menu with a TOC containing headers respecting the provided toc config", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-4/", "./tests/fixtures/menu-4/_site", {
        configPath: "./tests/fixtures/menu-4/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});

test("returns a menu with no TOC if it has been disabled via the config", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-5/", "./tests/fixtures/menu-5/_site", {
        configPath: "./tests/fixtures/menu-5/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});

test("works with a global data permalink", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-6/", "./tests/fixtures/menu-6/_site", {
        configPath: "./tests/fixtures/menu-6/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});

test("works with a directory data permalink", async () => {
    const elev = new Eleventy("./tests/fixtures/menu-7/", "./tests/fixtures/menu-7/_site", {
        configPath: "./tests/fixtures/menu-7/.eleventy.js",
    });

    const json = await elev.toJSON() as any[];
    const formattedResult = await formatHTML(json[0].content);

    expect(formattedResult).toMatchSnapshot();
});
