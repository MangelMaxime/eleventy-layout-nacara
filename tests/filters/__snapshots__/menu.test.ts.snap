// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns a menu with a TOC containing headers respecting the provided toc config 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1/index.html"
                    class="is-active"
                >
                    Page 1
                </a>
            </li>
            <li>
                <ul class="table-of-content">
                    <li data-toc-rank="3">
                        <a href="#Header-3-1" data-toc-element="true">
                            Header 3 - #1
                        </a>
                    </li>
                    <li data-toc-rank="4">
                        <a href="#Header-4-1" data-toc-element="true">
                            Header 4 - #1
                        </a>
                    </li>
                </ul>
            </li>
        </ul>
    </aside>
</div>
"
`;

exports[`returns a menu with a TOC containing only h2 headers if no toc data is provided 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1/index.html"
                    class="is-active"
                >
                    Page 1
                </a>
            </li>
            <li>
                <ul class="table-of-content">
                    <li data-toc-rank="2">
                        <a href="#Header-2-1" data-toc-element="true">
                            Header 2 - #1
                        </a>
                    </li>
                    <li data-toc-rank="2">
                        <a href="#Header-2-2" data-toc-element="true">
                            Header 2 - #2
                        </a>
                    </li>
                </ul>
            </li>
        </ul>
    </aside>
</div>
"
`;

exports[`returns a menu with no TOC if it has been disabled via the config 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li><a href="/Nacara/undefined" class="is-active">Page 1</a></li>
        </ul>
    </aside>
</div>
"
`;

exports[`returns nothing if there is not nacaraMenu provided 1`] = `""`;

exports[`supports nacaraMenu with a Section 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1/index.html"
                    class="is-active"
                >
                    Page 1
                </a>
            </li>
            <li>
                <a
                    href="/Nacara/docs/getting-started/page2/index.html"
                    class=""
                >
                    Page 2
                </a>
            </li>
        </ul>
    </aside>
</div>
"
`;

exports[`supports nacaraMenu with a Section 2`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1/index.html"
                    class=""
                >
                    Page 1
                </a>
            </li>
            <li>
                <a
                    href="/Nacara/docs/getting-started/page2/index.html"
                    class="is-active"
                >
                    Page 2
                </a>
            </li>
        </ul>
    </aside>
</div>
"
`;

exports[`works with a directory data permalink 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1.html"
                    class="is-active"
                >
                    Page 1
                </a>
            </li>
        </ul>
    </aside>
</div>
"
`;

exports[`works with a global data permalink 1`] = `
"<div class="menu-container">
    <aside class="menu">
        <p class="menu-label">Getting started</p>
        <ul class="menu-list">
            <li>
                <a
                    href="/Nacara/docs/getting-started/page1.html"
                    class="is-active"
                >
                    Page 1
                </a>
            </li>
        </ul>
    </aside>
</div>
"
`;
